packages/flutter/lib/src/rendering/auto_layout.dart

```diff
@@ -38,153 +37,154 @@ abstract class _AutoLayoutParamMixin {
-  /// active solver. If no implicit constraints needs to be applied, the object
-  /// may return null.
-  List<al.Constraint> _constructImplicitConstraints();
-
-  void _setupImplicitConstraints(al.Solver solver) {
-    List<al.Constraint> implicit = _constructImplicitConstraints();
+  List<al.Constraint> _implicitConstraints;
 
-    if (implicit == null || implicit.length == 0) {
+  void _addImplicitConstraints() {
+    assert(_renderBox != null);
+    if (_renderBox.parent == null)
       return;
-    }
-
-    al.Result result = solver.addConstraints(implicit);
+    assert(_renderBox.parent is RenderAutoLayout);
+    final RenderAutoLayout parent = _renderBox.parent;
+    final AutoLayoutParentData parentData = _renderBox.parentData;
+    final List<al.Constraint> implicit = parentData._constructImplicitConstraints();
+    if (implicit == null || implicit.isEmpty)
+      return;
+    final al.Result result = parent._solver.addConstraints(implicit);
     assert(result == al.Result.success);
-
+    parent.markNeedsLayout();
     _implicitConstraints = implicit;
   }
 
-  void _removeImplicitConstraints(al.Solver solver) {
-    if (_implicitConstraints == null || _implicitConstraints.length == 0) {
+  void _removeImplicitConstraints() {
+    assert(_renderBox != null);
+    if (_renderBox.parent == null)
       return;
-    }
-
-    al.Result result = solver.removeConstraints(_implicitConstraints);
+    if (_implicitConstraints == null || _implicitConstraints.isEmpty)
+      return;
+    assert(_renderBox.parent is RenderAutoLayout);
+    final RenderAutoLayout parent = _renderBox.parent;
+    final al.Result result = parent._solver.removeConstraints(_implicitConstraints);
     assert(result == al.Result.success);
-
+    parent.markNeedsLayout();
     _implicitConstraints = null;
   }
 }
 
-class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> with _AutoLayoutParamMixin {
-
-  AutoLayoutParentData(this._renderBox) {
-    _setupLayoutParameters(this);
-  }
+class AutoLayoutParentData extends ContainerBoxParentDataMixin<RenderBox> {
+  AutoLayoutParentData(this._renderBox);
 
   final RenderBox _renderBox;
 
-  void _applyAutolayoutParameterUpdates() {
-    // This is called by the parent's layout function
-    // to lay our box out.
-    assert(_renderBox.parentData == this);
-    assert(() {
-      final RenderAutoLayout parent = _renderBox.parent;
-      assert(parent.debugDoingThisLayout);
-    });
-    BoxConstraints size = new BoxConstraints.tightFor(
-      width: _rightEdge.value - _leftEdge.value,
-      height: _bottomEdge.value - _topEdge.value
+  AutoLayoutParams get params => _params;
+  AutoLayoutParams _params;
+  void set params(AutoLayoutParams value) {
+    if (_params == value)
+      return;
+    if (_params != null) {
+      _params._removeImplicitConstraints();
+      _params._renderBox = null;
+    }
+    _params = value;
+    if (_params != null) {
+      assert(_params._renderBox == null);
```

Comment:
This doesn't work quite as cleanly as I would like.  Ideally I would be able to associate a context with a parameter when I added to to a solver.  That way a single AutoLayoutParams object could interact with many RenderBoxes at once.  As currently implemented, this will blow up if you try to move the AutoLayoutParams object from one RenderBoxes to another and you add it to the new RenderBoxes before removing it from the old RenderBoxes.